### PR TITLE
ATM: Test for contradictory endpoint characteristics

### PR DESCRIPTION
**Pull request by tiferet**
Adds a test that surfaces endpoints with a set of characteristics that are logically incompatible with one another (e.g one high-confidence characteristic that implies a non-sink and another that implies a sink). If the test surfaces any such endpoints, this is a hint that some of our endpoint characteristics may be need to be adjusted.

Note that this test currently reveals several high-confidence negative characteristics that apply to some known sinks. We should experiment with lowering the confidence of these characteristics, so that they don't contribute negative training examples. These known exceptions are hard-coded in the test, similarly to `FilteredTruePositives.ql`, to make it easy to find all currently-known problems so we can fix them.

Closes https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2099

### CODE REVIEW

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ContradictoryEndpointCharacteristics.ql [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    confidence > characteristic.mediumConfidence() and
+    confidence2 > characteristic2.mediumConfidence()
+  ) and
+  errorMessage = "Endpoint has high-confidence positive and negative indicators for the same class"
+  or
+  // The endpoint's characteristics should not include indicators with confidence outside of [0, 1].
+  characteristic.appliesToEndpoint(endpoint) and
+  characteristic.hasImplications(_, _, confidence) and
+  (confidence < 0 or confidence > 1) and
+  errorMessage = "Endpoint has an indicator with confidence outside of [0, 1]"
```
Comment (adityasharad) [inline javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ContradictoryEndpointCharacteristics.ql:73]: I think this last test case can be phrased independently of the endpoint! You could make it a separate test case/query that looks for only the last 3 lines of this logic, without the `appliesToEndpoint(...)` call.
Slight advantages:
- you will catch discrepancies even if (for whatever reason) the test data doesn't actually contain an endpoint of the given characteristic
- slight speedup in this part of the test

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ContradictoryEndpointCharacteristics.ql [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      characteristic2 = ["DomBasedXssSink", "DOM"]
+      or
+      characteristic = ["DomBasedXssSink", "JQueryArgument"] and
+      characteristic2 = ["DomBasedXssSink", "JQueryArgument"]
+      or
+      characteristic = ["NosqlInjectionSink", "DatabaseAccess"] and
+      characteristic2 = ["NosqlInjectionSink", "DatabaseAccess"]
+      or
+      characteristic = ["SqlInjectionSink", "DatabaseAccess"] and
+      characteristic2 = ["SqlInjectionSink", "DatabaseAccess"]
```
Comment (adityasharad) [inline javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ContradictoryEndpointCharacteristics.ql:52]: Pull this logic out into its own helper predicate, e.g. `knownCompatibleCharacteristics(c1, c2)`.

**Review by adityasharad — COMMENTED** [review]
Neat. Couple of suggestions.

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ContradictoryEndpointCharacteristics.ql (hunk 2 above)
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ContradictoryEndpointCharacteristics.ql]: 👍 done in cf29cde2e8377a0d450267831d343b42b72c8b21

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ContradictoryEndpointCharacteristics.ql (hunk 1 above)
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ContradictoryEndpointCharacteristics.ql]: Good idea! Done in cf29cde2e8377a0d450267831d343b42b72c8b21

**Pull request merged**